File: May-LeetCoding-Challenge/First-Bad-Version/First-Bad-Version.scala
class Solution extends VersionControl {
  def firstBadVersion(n: Int): Int = firstBadVersionFromRange(1, n)

  def firstBadVersionFromRange(start: Int, end: Int): Int =
    if (start == end) start
    else {
      val mid = start + (end - start)/2
      if (isBadVersion(mid)) firstBadVersionFromRange(start, mid)
      else firstBadVersionFromRange(mid + 1, end)
    }
}

File: May-LeetCoding-Challenge/First-Bad-Version/First-Bad-Version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int start = 1, end = n;
        int mid = start + (end - start) / 2;
        while (start < end) {
            if (isBadVersion(mid)) {
                end = mid;
            } else {
                start = mid + 1;
            }
            mid = start + (end - start) / 2;
        }
        return mid;
    }
}
